### PR TITLE
BUG: DataFrame.stack sometimes sorting the resulting index

### LINKED ISSUE #53824: BUG: DataFrame.stack sorting index values in rare cases
```
labels = list("zyx")
data = [14, 15, 12, 13, 10, 11]
ordered = False
cidx = pd.CategoricalIndex(labels, categories=sorted(labels), ordered=ordered)
cidx2 = pd.CategoricalIndex(["u", "v"], ordered=ordered)
midx = MultiIndex.from_product([cidx, cidx2])
df = DataFrame([sorted(data)], columns=midx)

print(df)
#     z       y       x    
#     u   v   u   v   u   v
# 0  10  11  12  13  14  15

print(df.stack([0, 1]))
# 0  x  u    14
#       v    15
#    y  u    12
#       v    13
#    z  u    10
#       v    11
# dtype: int64
```

The stacked result should maintain the order `z y x`.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #53824 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -498,7 +498,8 @@ Reshaping
 - Bug in :meth:`DataFrame.idxmin` and :meth:`DataFrame.idxmax`, where the axis dtype would be lost for empty frames (:issue:`53265`)
 - Bug in :meth:`DataFrame.merge` not merging correctly when having ``MultiIndex`` with single level (:issue:`52331`)
 - Bug in :meth:`DataFrame.stack` losing extension dtypes when columns is a :class:`MultiIndex` and frame contains mixed dtypes (:issue:`45740`)
-- Bug in :meth:`DataFrame.stack` sorting columns lexicographically (:issue:`53786`)
+- Bug in :meth:`DataFrame.stack` sorting columns lexicographically in rare cases (:issue:`53786`)
+- Bug in :meth:`DataFrame.stack` sorting index lexicographically in rare cases (:issue:`53824`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.0.rst:502]: There are tons of tests for stacking not sorting the order; only one of them is impacted by this bug. I haven't been able to figure out a way to describe the circumstances this happens under.

**Comment by rhshadrach** [comment]
@mroeschke - this removes all uses of `sort` in DataFrame.stack. Looking at #15105 again, it appear to me the cause of all the issues was the `unstack` behavior. #53282 adding `sort=True/False` to unstack resolves that issue alone, and the use of this argument controls the sorting of the values.

On the other hand, #53282 does not modify the sorting of the values (values here being the levels + codes combination), but rather just the sorting of the levels (the codes are then also adjusted so the values come out the same). So I think we can remove the `sort` argument from stack and still resolve #15105 with this PR. Would you be okay doing this here?

**Comment by mroeschke** [comment]
> So I think we can remove the sort argument from stack and still resolve https://github.com/pandas-dev/pandas/issues/15105 with this PR. Would you be okay with me doing this here?

Hmm it would be nice for API consistency for `stack/unstack` to both have `sort` keywords. Would it be difficult to implement `sort=True` for `stack`?

**Comment by rhshadrach** [comment]
> Would it be difficult to implement `sort=True` for `stack`?

No - it's not. But I was leaning the other way: remove `sort` from unstack after changing it's default to False. To sort, it would just be a call to `sort_index(axis=1)` after. There is a perf impact here:

```
arrays = [np.arange(100).repeat(100), np.roll(np.tile(np.arange(100), 100), 25)]
index = MultiIndex.from_arrays(arrays)
df = DataFrame(np.random.randn(10000, 4), index=index)

%timeit df.unstack(1, sort=True)
602 µs ± 5.38 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit df.unstack(1, sort=False)
638 µs ± 2.78 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit df.unstack(1, sort=False).sort_index(axis=1)
1.04 ms ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

In general I advocate for not having arguments when it's just an additional call, but perhaps the perf benefit is worth it here?

**Comment by mroeschke** [comment]
> In general I advocate for not having arguments when it's just an additional call, but perhaps the perf benefit is worth it here?

Yeah I agree with not having arguments when its an additional call, additionally I think the `sort=False` behavior is a more natural default behavior so I wouldn't mind moving toward an eventual removal of the keyword where we do not sort by default

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Nice thanks @rhshadrach

**Pull request merged**